In [1]:
import tensorflow as tf

C:\Anaconda3\lib\site-packages\h5py\__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters


# Linear Regression
★ **Writing Mathematic 참고 페이지**: [http://csrgxtu.github.io/2015/03/20/Writing-Mathematic-Fomulars-in-Markdown/](http://csrgxtu.github.io/2015/03/20/Writing-Mathematic-Fomulars-in-Markdown/)


```
(예시)
어떤 학생이 몇시간 공부했더니 얼마 정도의 성적이 나오더라
```
* supervised learning
> **x, y의 데이터를 가지고** 학습을 시킨다.

최종적인 목표는 0 ~ 100점 사이의 스코어 **범위를 예측**한다. 

* regression
> 범위를 예측시키는것은, 이런 형태의 머신러닝을 supervised learning중에서도 regression이라고 한다.

* training
> 이미 나온 데이터를 가지고 학습을 시키는 과정을 training이라고 하며, 그 training에 쓰이는 데이터를 training data라고 한다. 

* regression모델
> regression모델이 training data를 가지고 학습을 하면, 
어떤 모델을 만든다. 모델이 만들어졌다는 것은 학습이 끝났다는 것을 뜻한다.

* regression을 사용한다는 것?
> "어떤 학생이 7시간 공부했는데 몇점이나 받을 수 있겠니?" 하고 기계한테 물어보면, regression은 학습된 것으로 기반으로 리턴값(y)를 출력해준다.


\* **[회귀분석(regression)](https://ko.wikipedia.org/wiki/%ED%9A%8C%EA%B7%80%EB%B6%84%EC%84%9D): ** 통계학에서, 회귀분석(回歸分析, 영어: regression analysis)은 <u>관찰된 연속형 변수들에 대해 두 변수 사이의 모형을 구한뒤 적합도를 측정해 내는 분석 방법</u>이다.

* regression 모델을 학습한다는 것?
> <u>**하나의 가설을 세울 필요가 있다.**</u> <br>
"우리 데이터에 대해서 잘 모르겠지만, **이런 형태로 맞을꺼야!!**"
>> Linear(라는 이름이 붙여진다면,)한 모델이 우리 데이터에 맞을꺼다.

\* **[선형(linear)](https://ko.wikipedia.org/wiki/%EC%84%A0%ED%98%95%EC%84%B1): **  <u> **직선처럼** 똑바른 도형, 또는 그와 비슷한 성질을 갖는 대상</u>이라는 뜻으로, 이러한 성질을 갖고 있는 변환 등에 대하여 쓰는 용어이다. **함수의 경우**, <u>어떠한 함수가 진행하는 모양이 '직선'이라는 의미로 사용</u>된다. 이러한 개념은 수학, 물리학 등에서 많이 사용된다. 다른 말로 1차(一次)라고도 한다. (단어 '1차' 자체는, '선형'을 의미하지 않는 경우도 많다.)

```
(예시)
공부를 많이하면 할수록, 성적이 올라간다. 
```

**Linear로 설명할 수 있는, Linear로 모델을 세울 수 있는 경우가 많다. **

**Linear하게 가설을 세운다는 것은, <u>어떤 데이터가 있다면, 여기에 잘 맞는 Linear한 선을 찾는다는 것이라고 할 수 있다.</u>**<br>
**데이터에 알맞는 Linear한 선을 찾는 것이** 학습을 하는 것이다.

<br>
따라해보자: [Simple AI/ Linear Regression](https://medium.com/simple-ai/linear-regression-intro-to-machine-learning-6-6e320dbdaf06) <br>
따라해보자: [머신러닝스터디 4. 회귀분석](https://medium.com/mathpresso/mathpresso-%EB%A8%B8%EC%8B%A0-%EB%9F%AC%EB%8B%9D-%EC%8A%A4%ED%84%B0%EB%94%94-4-%ED%9A%8C%EA%B7%80-%EB%B6%84%EC%84%9D-regression-1-6d6cc0aaa483) <br>
참고: [김성훈의 교수님의 강의가 정리되어있는 블로그](http://pythonkim.tistory.com/7)<br>
참고: [수학포기자를 위한 딥러닝 로지스틱 회귀를 이용한 분류 모델](http://bcho.tistory.com/1142)<br>
라이브러리: [python/scikit/linearregression](http://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LinearRegression.html)<br>
참고: [ML 01:LINIREAR REGRESSION](http://1ambda.github.io/data-analysis/machine-learning-week-1/)


# (Linear) Hypothesis
\\(H(x) = Wx + b\\) <br>
**주어진 x값에 대해서 예측을 어떻게 할 것인가가 Hypothesis**
\\(h_{\Theta}x = \Theta_{0} + \Theta_{1}x\\)

<img src=http://mercris.files.wordpress.com/2012/07/screen-shot-2012-07-17-at-2-12-05-pm.png?w=584>
[https://mercris.files.wordpress.com/2012/07/screen-shot-2012-07-17-at-2-12-05-pm.png?w=584](https://mercris.files.wordpress.com/2012/07/screen-shot-2012-07-17-at-2-12-05-pm.png?w=584)

* Which hypothesis is better?
여러개 그어진 linear한 선 중에서, **어떤 선이 우리가 가지고 있는 데이터와 <u>가장 가까운 선인가?</u>** 

* cost function, loss function, 거리로 측정
실제 데이터와 가설이 나타내는 데이터 점의 거리가 멀면 나쁜것, 거기가 가까우면 좋은것
> 우리가 세운 가설과 실제 데이터가 얼마나 다른가를 나타낸다.<br>
cost function->  \\({(H(x)-y)}^{2}\\)
<br><br>
normal하게 표현하면, <br>

\\(cost(W, b)\\) = \\(\frac{1}{m}\sum_{i=1}^{m} (H(x^i)-y^i)^2\\) <br>
<br>
\\(J(\Theta_{0}, \Theta_{1})\\) =

<img src="http://s0.wp.com/latex.php?latex=J%28%5Ctheta_%7B0%7D%2C+%5Ctheta_%7B1%7D%29+%3D+%5Cdfrac+%7B1%7D%7B2m%7D+%5Csum+%5Climits_%7Bi%3D1%7D%5E%7Bm%7D+%28h_%7B%5Ctheta%7D+%28x%5E%7B%28i%29%7D%29+-+y%5E%7B%28i%29%7D%29%5E2&bg=ffffff&fg=333333&s=0">

[http://www.holehouse.org/](http://www.holehouse.org/](http://www.holehouse.org/](http://www.holehouse.org/)
<br>
```
이 식을 cost function 또는 squred error function 이라 부른다. 여기서 1/2m 으로 나누는 이유에 대해 좀 궁금해서 구글링 해봤는데, 1/m 으로 나누는 이유는 squared error 에 대해 mean 을 얻기 위한거고, 1/2 로 다시 나누는 이유는 미분했을때 나오는 2 를 제거하기 위해서다.
```
출저: [http://1ambda.github.io/data-analysis/machine-learning-week-1/](http://1ambda.github.io/data-analysis/machine-learning-week-1/)

##### W, b가 달라질 수록 cost function은 커질수도 작아질 수 있다. 즉, W, b를 조정해서 cost function을 가장 작은 사이즈로 조정하는 것을 `학습`이라고 한다.

## 1. graph 만들기

### \\(H(x) = Wx + b\\) <br>
variable로 정의할 수 있다. <br>
여기서 variable은 기존의 프로그램에서의 변수와 다른 의미이다. <br>
* **텐서플로우가 사용하는 Variable = tensorflow가 자체적으로 변경시키는 값 = trainable한 Variable**
* 값을 모르니까 random값으로 준다.
* tf.random_normal([**shape**])

In [4]:
x_train = [1, 2, 3]
y_train = [1, 2, 3]

W = tf.Variable(tf.random_normal([1]), name='weight')
b = tf.Variable(tf.random_normal([1]), name='bias')
hypothesis = x_train * W + b # node

### \\(cost(W, b)\\) = \\(\frac{1}{m}\sum_{i=1}^{m} (H(x^i)-y^i)^2\\)

In [11]:
t = [1., 2., 3., 4.]
print(tf.reduce_mean(t))

Tensor("Mean_2:0", shape=(), dtype=float32)


In [7]:
cost = tf.reduce_mean(tf.square(hypothesis - y_train))
# 평균을 내주는 부분

### GradientDescent

In [8]:
optimizer = tf.train.GradientDescentOptimizer(learning_rate=0.01)
# optimizer 함수 호출
train = optimizer.minimize(cost)
# minimize

## 2. session만들기

In [12]:
sess = tf.Session()
# Variable을 사용하기 전에는 global.. 
sess.run(tf.global_variables_initializer())

**minimize하는 작업, 노드**를 train이라고 하였다.<br>
train을 시키면 연결연결되어서 W,b로 연결된다.

In [13]:
for step in range(2001):
    sess.run(train)
    if step % 20 ==0:
        print(step, sess.run(cost), sess.run(W), sess.run(b))

0 17.579319 [-1.2386743] [0.70399326]
20 0.41764104 [0.25290775] [1.2807759]
40 0.23833752 [0.4200277] [1.2786555]
60 0.21518372 [0.45985416] [1.2240926]
80 0.19542176 [0.48643598] [1.1670913]
100 0.1774851 [0.51068556] [1.1122925]
120 0.1611948 [0.53369236] [1.0600238]
140 0.14639966 [0.5556081] [1.0102069]
160 0.13296251 [0.576493] [0.96273094]
180 0.120758675 [0.59639627] [0.9174861]
200 0.10967496 [0.6153642] [0.8743676]
220 0.09960855 [0.6334407] [0.8332755]
240 0.09046611 [0.6506676] [0.79411465]
260 0.082162715 [0.66708493] [0.7567941]
280 0.074621536 [0.6827308] [0.7212277]
300 0.06777248 [0.69764113] [0.6873327]
320 0.06155205 [0.71185094] [0.65503055]
340 0.0559026 [0.7253928] [0.6242467]
360 0.050771613 [0.73829836] [0.59490937]
380 0.04611164 [0.75059736] [0.56695086]
400 0.041879326 [0.7623183] [0.5403063]
420 0.03803545 [0.7734885] [0.5149139]
440 0.034544438 [0.7841337] [0.4907149]
460 0.03137381 [0.79427856] [0.46765316]
480 0.028494185 [0.80394673] [0.44567516]
500 0.0

![Tensor image](tensorimg.png "tensor")
train만 실행시킨다는 의미는 **그래프를 따라 들어가서 W, b의 값을 조정할 수 있다는 의미** <br>
이때 학습이 일어나게 되고, cost와 W, b의 값이 어떻게되는지 볼 수 있다. 

# Placholders

In [61]:
# Lab 2 Linear Regression
import tensorflow as tf
tf.set_random_seed(777)  # for reproducibility

# Try to find values for W and b to compute y_data = W * x_data + b
# We know that W should be 1 and b should be 0
# But let's use TensorFlow to figure it out
W = tf.Variable(tf.random_normal([1]), name='weight')
b = tf.Variable(tf.random_normal([1]), name='bias')

# Now we can use X and Y in place of x_data and y_data
# # placeholders for a tensor that will be always fed using feed_dict
# See http://stackoverflow.com/questions/36693740/
X = tf.placeholder(tf.float32, shape=[None])
Y = tf.placeholder(tf.float32, shape=[None])

# Our hypothesis XW+b
hypothesis = X * W + b

# cost/loss function
cost = tf.reduce_mean(tf.square(hypothesis - Y))

# Minimize
optimizer = tf.train.GradientDescentOptimizer(learning_rate=0.05)
train = optimizer.minimize(cost)

# Launch the graph in a session.
sess = tf.Session()
# Initializes global variables in the graph.
sess.run(tf.global_variables_initializer())

print("=================================================")
# Fit the line
for step in range(2001):
    cost_val, W_val, b_val, _ = \
        sess.run([cost, W, b, train],
                 feed_dict={X: [1, 2, 3], Y: [1, 2, 3]})
    if step % 20 == 0:
        print(step, cost_val, W_val, b_val)

# Learns best fit W:[ 1.],  b:[ 0]


# Testing our model
print(sess.run(hypothesis, feed_dict={X: [5]}))
print(sess.run(hypothesis, feed_dict={X: [2.5]}))
print(sess.run(hypothesis, feed_dict={X: [1.5, 3.5]}))


print("===============================================")
# Fit the line with new training data
for step in range(2001):
    cost_val, W_val, b_val, _ = \
        sess.run([cost, W, b, train],
                 feed_dict={X: [1, 2, 3, 4, 5],
                            Y: [2.1, 3.1, 4.1, 5.1, 6.1]})
    if step % 20 == 0:
        print(step, cost_val, W_val, b_val)

# Testing our model
print(sess.run(hypothesis, feed_dict={X: [5]}))
print(sess.run(hypothesis, feed_dict={X: [2.5]}))
print(sess.run(hypothesis, feed_dict={X: [1.5, 3.5]}))

0 7.2130203 [1.2073979] [-1.703354]
20 0.20484729 [1.519349] [-1.1806024]
40 0.12628807 [1.4077796] [-0.92697877]
60 0.0778564 [1.3201779] [-0.7278398]
80 0.047998365 [1.2513956] [-0.5714811]
100 0.02959094 [1.1973892] [-0.44871226]
120 0.018242776 [1.1549851] [-0.3523174]
140 0.011246641 [1.1216903] [-0.27663052]
160 0.0069335364 [1.095548] [-0.2172032]
180 0.0042745206 [1.0750219] [-0.17054224]
200 0.0026352338 [1.0589052] [-0.13390543]
220 0.0016246252 [1.0462509] [-0.10513914]
240 0.0010015763 [1.0363151] [-0.08255263]
260 0.0006174725 [1.0285137] [-0.06481826]
280 0.00038067004 [1.0223882] [-0.05089362]
300 0.00023468293 [1.0175786] [-0.03996036]
320 0.00014468103 [1.0138023] [-0.03137582]
340 8.919481e-05 [1.0108372] [-0.02463547]
360 5.4988388e-05 [1.008509] [-0.01934314]
380 3.3900335e-05 [1.0066811] [-0.01518776]
400 2.0900074e-05 [1.0052458] [-0.01192507]
420 1.2885536e-05 [1.0041189] [-0.00936335]
440 7.943765e-06 [1.0032341] [-0.00735188]
460 4.897369e-06 [1.0025394] [-0.00

1860 2.2851055e-12 [1.0000011] [1.0999962]
1880 2.2851055e-12 [1.0000011] [1.0999962]
1900 2.2851055e-12 [1.0000011] [1.0999962]
1920 2.2851055e-12 [1.0000011] [1.0999962]
1940 2.2851055e-12 [1.0000011] [1.0999962]
1960 2.2851055e-12 [1.0000011] [1.0999962]
1980 2.2851055e-12 [1.0000011] [1.0999962]
2000 2.2851055e-12 [1.0000011] [1.0999962]
[6.1000013]
[3.599999]
[2.599998 4.6     ]


In [34]:
W = tf.Variable(tf.random_normal([1]), name='weight')
b = tf.Variable(tf.random_normal([1]), name='bias')

X = tf.placeholder(tf.float32, shape=[None])
Y = tf.placeholder(tf.float32, shape=[None])

hypothesis = X * W + b
cost = tf.reduce_mean(tf.square(hypothesis - Y))
optimizer = tf.train.GradientDescentOptimizer(learning_rate=0.01)
train = optimizer.minimize(cost)

sess = tf.Session()
sess.run(tf.global_variables_initializer())

for step in range(2001):
    cost_val, W_val, b_val, _ = \
        sess.run([cost, W, b, train], 
                 feed_dict={X: [1, 2, 3], Y: [1, 2, 3]})
    if step % 20 == 0:
        print("==========step, cost_val, W_val, b_val=====")
        print(step, cost_val, W_val, b_val)

print("=========feed_dict={X: [5]}=============")
print(sess.run(hypothesis, feed_dict={X: [5]}))
print("=========feed_dict={X: [2.5]}=============")
print(sess.run(hypothesis, feed_dict={X: [2.5]}))
print("=========feed_dict={X: [1.5, 3.5]}))=============")
print(sess.run(hypothesis, feed_dict={X: [1.5, 3.5]}))

==========step, cost_val, W_val, b_val=====
0 13.288369 [-0.29553396] [-0.4724274]
==========step, cost_val, W_val, b_val=====
20 0.121346466 [0.84582084] [0.02507534]
==========step, cost_val, W_val, b_val=====
40 0.0018854747 [0.95594823] [0.06915468]
==========step, cost_val, W_val, b_val=====
60 0.0007302738 [0.96781486] [0.07021407]
==========step, cost_val, W_val, b_val=====
80 0.0006543403 [0.9702602] [0.06732462]
==========step, cost_val, W_val, b_val=====
100 0.0005942019 [0.9717466] [0.06419967]
==========step, cost_val, W_val, b_val=====
120 0.0005396614 [0.97308296] [0.06118624]
==========step, cost_val, W_val, b_val=====
140 0.0004901315 [0.9743487] [0.05831107]
==========step, cost_val, W_val, b_val=====
160 0.00044514347 [0.97555435] [0.05557071]
==========step, cost_val, W_val, b_val=====
180 0.00040428768 [0.97670317] [0.0529591]
==========step, cost_val, W_val, b_val=====
200 0.0003671817 [0.97779804] [0.05047024]
==========step, cost_val, W_val, b_val=====
220 0.0003

In [62]:
X = tf.placeholder(tf.float32)
Y = tf.placeholder(tf.float32)

hypothesis = X * W + b
cost = tf.reduce_mean(tf.square(hypothesis - Y))
optimizer = tf.train.GradientDescentOptimizer(learning_rate=0.15)
train = optimizer.minimize(cost)
sess = tf.Session()
sess.run(tf.global_variables_initializer())

for step in range(2001):
    cost_val, W_val, b_val, _ = \
        sess.run([cost, W, b, train], 
                 feed_dict={X: [1, 2, 3], Y: [1, 2, 3]})
    if step % 20 == 0:
        print("==========step, cost_val, W_val, b_val=====")
        print(step, cost_val, W_val, b_val)

print("=========feed_dict={X: [5]}=============")
print(sess.run(hypothesis, feed_dict={X: [5]}))
print("=========feed_dict={X: [2.5]}=============")
print(sess.run(hypothesis, feed_dict={X: [2.5]}))
print("=========feed_dict={X: [1.5, 3.5]}))=============")
print(sess.run(hypothesis, feed_dict={X: [1.5, 3.5]}))

==========step, cost_val, W_val, b_val=====
0 7.2130203 [2.3222513] [-1.1692626]
==========step, cost_val, W_val, b_val=====
20 0.076475 [1.3097904] [-0.7037172]
==========step, cost_val, W_val, b_val=====
40 0.017600907 [1.1485301] [-0.33764347]
==========step, cost_val, W_val, b_val=====
60 0.0040509207 [1.0712563] [-0.16198236]
==========step, cost_val, W_val, b_val=====
80 0.0009323374 [1.0341847] [-0.07771005]
==========step, cost_val, W_val, b_val=====
100 0.00021457928 [1.0164] [-0.03728088]
==========step, cost_val, W_val, b_val=====
120 4.938722e-05 [1.0078677] [-0.01788535]
==========step, cost_val, W_val, b_val=====
140 1.1366555e-05 [1.0037746] [-0.00858039]
==========step, cost_val, W_val, b_val=====
160 2.6159598e-06 [1.0018109] [-0.00411647]
==========step, cost_val, W_val, b_val=====
180 6.021246e-07 [1.0008688] [-0.00197485]
==========step, cost_val, W_val, b_val=====
200 1.3858663e-07 [1.0004168] [-0.00094742]
==========step, cost_val, W_val, b_val=====
220 3.1919594e

==========step, cost_val, W_val, b_val=====
1840 3.7895614e-14 [1.0000004] [-6.212066e-07]
==========step, cost_val, W_val, b_val=====
1860 3.7895614e-14 [1.0000004] [-6.212066e-07]
==========step, cost_val, W_val, b_val=====
1880 3.7895614e-14 [1.0000004] [-6.212066e-07]
==========step, cost_val, W_val, b_val=====
1900 3.7895614e-14 [1.0000004] [-6.212066e-07]
==========step, cost_val, W_val, b_val=====
1920 3.7895614e-14 [1.0000004] [-6.212066e-07]
==========step, cost_val, W_val, b_val=====
1940 3.7895614e-14 [1.0000004] [-6.212066e-07]
==========step, cost_val, W_val, b_val=====
1960 3.7895614e-14 [1.0000004] [-6.212066e-07]
==========step, cost_val, W_val, b_val=====
1980 3.7895614e-14 [1.0000004] [-6.212066e-07]
==========step, cost_val, W_val, b_val=====
2000 3.7895614e-14 [1.0000004] [-6.212066e-07]
=========feed_dict={X: [5]}=============
[5.0000014]
=========feed_dict={X: [2.5]}=============
[2.5000002]
=========feed_dict={X: [1.5, 3.5]}))=============
[1.4999999 3.5000005]


#### step = 20
```
step = 20
==========step, cost_val, W_val, b_val=====
0 13.288369 [-0.29553396] [-0.4724274]
==========step, cost_val, W_val, b_val=====
20 0.121346466 [0.84582084] [0.02507534]
==========step, cost_val, W_val, b_val=====
40 0.0018854747 [0.95594823] [0.06915468]
==========step, cost_val, W_val, b_val=====
60 0.0007302738 [0.96781486] [0.07021407]
==========step, cost_val, W_val, b_val=====
80 0.0006543403 [0.9702602] [0.06732462]
==========step, cost_val, W_val, b_val=====
100 0.0005942019 [0.9717466] [0.06419967]
==========step, cost_val, W_val, b_val=====
120 0.0005396614 [0.97308296] [0.06118624]
==========step, cost_val, W_val, b_val=====
140 0.0004901315 [0.9743487] [0.05831107]
==========step, cost_val, W_val, b_val=====
160 0.00044514347 [0.97555435] [0.05557071]
==========step, cost_val, W_val, b_val=====
180 0.00040428768 [0.97670317] [0.0529591]
==========step, cost_val, W_val, b_val=====
200 0.0003671817 [0.97779804] [0.05047024]
==========step, cost_val, W_val, b_val=====
220 0.00033348193 [0.9788414] [0.04809835]
==========step, cost_val, W_val, b_val=====
240 0.00030287384 [0.9798356] [0.04583799]
==========step, cost_val, W_val, b_val=====
260 0.00027507325 [0.9807835] [0.04368376]
==========step, cost_val, W_val, b_val=====
280 0.00024982673 [0.9816866] [0.04163074]
==========step, cost_val, W_val, b_val=====
300 0.00022689575 [0.9825473] [0.03967422]
==========step, cost_val, W_val, b_val=====
320 0.00020607124 [0.98336744] [0.03780967]
==========step, cost_val, W_val, b_val=====
340 0.00018715672 [0.98414916] [0.03603275]
==========step, cost_val, W_val, b_val=====
360 0.00016997811 [0.98489416] [0.03433931]
==========step, cost_val, W_val, b_val=====
380 0.00015437823 [0.98560405] [0.03272545]
==========step, cost_val, W_val, b_val=====
400 0.00014020682 [0.98628056] [0.03118748]
==========step, cost_val, W_val, b_val=====
420 0.00012733859 [0.98692536] [0.02972179]
==========step, cost_val, W_val, b_val=====
440 0.00011564984 [0.9875398] [0.02832496]
==========step, cost_val, W_val, b_val=====
460 0.00010503468 [0.9881254] [0.02699377]
==========step, cost_val, W_val, b_val=====
480 9.5395946e-05 [0.98868346] [0.02572516]
==========step, cost_val, W_val, b_val=====
500 8.664e-05 [0.9892152] [0.0245162]
==========step, cost_val, W_val, b_val=====
520 7.868841e-05 [0.98972213] [0.02336403]
==========step, cost_val, W_val, b_val=====
540 7.146429e-05 [0.9902053] [0.02226595]
==========step, cost_val, W_val, b_val=====
560 6.490495e-05 [0.99066556] [0.02121948]
==========step, cost_val, W_val, b_val=====
580 5.894755e-05 [0.99110425] [0.0202222]
==========step, cost_val, W_val, b_val=====
600 5.3536685e-05 [0.9915223] [0.01927179]
==========step, cost_val, W_val, b_val=====
620 4.8622867e-05 [0.9919207] [0.0183661]
==========step, cost_val, W_val, b_val=====
640 4.4159886e-05 [0.99230045] [0.01750294]
==========step, cost_val, W_val, b_val=====
660 4.0107276e-05 [0.9926623] [0.01668036]
==========step, cost_val, W_val, b_val=====
680 3.6425732e-05 [0.9930071] [0.01589645]
==========step, cost_val, W_val, b_val=====
700 3.3083106e-05 [0.9933357] [0.01514938]
==========step, cost_val, W_val, b_val=====
720 3.0046698e-05 [0.99364895] [0.01443741]
==========step, cost_val, W_val, b_val=====
740 2.728796e-05 [0.99394745] [0.01375891]
==========step, cost_val, W_val, b_val=====
760 2.4783643e-05 [0.9942319] [0.01311229]
==========step, cost_val, W_val, b_val=====
780 2.2508806e-05 [0.9945029] [0.01249607]
==========step, cost_val, W_val, b_val=====
800 2.0442982e-05 [0.9947612] [0.01190883]
==========step, cost_val, W_val, b_val=====
820 1.856673e-05 [0.9950074] [0.01134921]
==========step, cost_val, W_val, b_val=====
840 1.6862628e-05 [0.9952421] [0.01081584]
==========step, cost_val, W_val, b_val=====
860 1.5315005e-05 [0.9954657] [0.01030753]
==========step, cost_val, W_val, b_val=====
880 1.3909636e-05 [0.9956788] [0.00982311]
==========step, cost_val, W_val, b_val=====
900 1.263285e-05 [0.99588186] [0.00936148]
==========step, cost_val, W_val, b_val=====
920 1.1473053e-05 [0.9960754] [0.00892153]
==========step, cost_val, W_val, b_val=====
940 1.0420663e-05 [0.9962598] [0.00850227]
==========step, cost_val, W_val, b_val=====
960 9.464226e-06 [0.9964356] [0.00810269]
==========step, cost_val, W_val, b_val=====
980 8.595252e-06 [0.9966031] [0.00772192]
==========step, cost_val, W_val, b_val=====
1000 7.8066205e-06 [0.9967627] [0.00735905]
==========step, cost_val, W_val, b_val=====
1020 7.0901237e-06 [0.99691486] [0.00701323]
==========step, cost_val, W_val, b_val=====
1040 6.4390806e-06 [0.9970598] [0.00668364]
==========step, cost_val, W_val, b_val=====
1060 5.8483315e-06 [0.997198] [0.00636956]
==========step, cost_val, W_val, b_val=====
1080 5.311618e-06 [0.9973297] [0.00607024]
==========step, cost_val, W_val, b_val=====
1100 4.8240404e-06 [0.9974552] [0.00578496]
==========step, cost_val, W_val, b_val=====
1120 4.3813016e-06 [0.99757475] [0.00551309]
==========step, cost_val, W_val, b_val=====
1140 3.979032e-06 [0.9976888] [0.005254]
==========step, cost_val, W_val, b_val=====
1160 3.6137535e-06 [0.99779737] [0.00500707]
==========step, cost_val, W_val, b_val=====
1180 3.282186e-06 [0.9979009] [0.00477175]
==========step, cost_val, W_val, b_val=====
1200 2.9808982e-06 [0.9979995] [0.00454751]
==========step, cost_val, W_val, b_val=====
1220 2.7072429e-06 [0.99809355] [0.0043338]
==========step, cost_val, W_val, b_val=====
1240 2.4587478e-06 [0.99818313] [0.00413012]
==========step, cost_val, W_val, b_val=====
1260 2.2331108e-06 [0.9982685] [0.00393603]
==========step, cost_val, W_val, b_val=====
1280 2.0283114e-06 [0.9983499] [0.00375107]
==========step, cost_val, W_val, b_val=====
1300 1.8419613e-06 [0.99842745] [0.00357479]
==========step, cost_val, W_val, b_val=====
1320 1.6729592e-06 [0.99850136] [0.0034068]
==========step, cost_val, W_val, b_val=====
1340 1.5194445e-06 [0.99857175] [0.00324671]
==========step, cost_val, W_val, b_val=====
1360 1.3800912e-06 [0.99863887] [0.00309413]
==========step, cost_val, W_val, b_val=====
1380 1.2532952e-06 [0.9987028] [0.00294873]
==========step, cost_val, W_val, b_val=====
1400 1.138403e-06 [0.99876374] [0.00281018]
==========step, cost_val, W_val, b_val=====
1420 1.0338407e-06 [0.99882185] [0.00267813]
==========step, cost_val, W_val, b_val=====
1440 9.389278e-07 [0.99887717] [0.0025523]
==========step, cost_val, W_val, b_val=====
1460 8.527798e-07 [0.99893] [0.00243239]
==========step, cost_val, W_val, b_val=====
1480 7.7450903e-07 [0.9989802] [0.0023181]
==========step, cost_val, W_val, b_val=====
1500 7.035908e-07 [0.9990281] [0.00220917]
==========step, cost_val, W_val, b_val=====
1520 6.3886506e-07 [0.99907386] [0.00210538]
==========step, cost_val, W_val, b_val=====
1540 5.8032634e-07 [0.9991174] [0.00200644]
==========step, cost_val, W_val, b_val=====
1560 5.2704655e-07 [0.9991588] [0.00191214]
==========step, cost_val, W_val, b_val=====
1580 4.7867485e-07 [0.9991984] [0.00182231]
==========step, cost_val, W_val, b_val=====
1600 4.347643e-07 [0.999236] [0.00173669]
==========step, cost_val, W_val, b_val=====
1620 3.948202e-07 [0.9992719] [0.0016551]
==========step, cost_val, W_val, b_val=====
1640 3.5860717e-07 [0.999306] [0.00157735]
==========step, cost_val, W_val, b_val=====
1660 3.2582957e-07 [0.9993387] [0.00150326]
==========step, cost_val, W_val, b_val=====
1680 2.9585397e-07 [0.9993698] [0.00143263]
==========step, cost_val, W_val, b_val=====
1700 2.6871479e-07 [0.9993993] [0.00136533]
==========step, cost_val, W_val, b_val=====
1720 2.4409027e-07 [0.9994275] [0.00130121]
==========step, cost_val, W_val, b_val=====
1740 2.2163562e-07 [0.99945444] [0.00124009]
==========step, cost_val, W_val, b_val=====
1760 2.0133223e-07 [0.99948] [0.00118183]
==========step, cost_val, W_val, b_val=====
1780 1.8286438e-07 [0.9995044] [0.00112634]
==========step, cost_val, W_val, b_val=====
1800 1.6611837e-07 [0.9995277] [0.00107345]
==========step, cost_val, W_val, b_val=====
1820 1.5088625e-07 [0.99954987] [0.00102304]
==========step, cost_val, W_val, b_val=====
1840 1.3703813e-07 [0.999571] [0.000975]
==========step, cost_val, W_val, b_val=====
1860 1.2449529e-07 [0.99959123] [0.00092924]
==========step, cost_val, W_val, b_val=====
1880 1.1307293e-07 [0.99961036] [0.00088563]
==========step, cost_val, W_val, b_val=====
1900 1.0265675e-07 [0.9996286] [0.00084404]
==========step, cost_val, W_val, b_val=====
1920 9.3276185e-08 [0.99964607] [0.0008044]
==========step, cost_val, W_val, b_val=====
1940 8.47452e-08 [0.99966276] [0.00076666]
==========step, cost_val, W_val, b_val=====
1960 7.69675e-08 [0.99967843] [0.00073068]
==========step, cost_val, W_val, b_val=====
1980 6.9909824e-08 [0.99969363] [0.00069637]
==========step, cost_val, W_val, b_val=====
2000 6.349562e-08 [0.99970794] [0.00066373]
=========feed_dict={X: [5]}=============
[4.9992037]
=========feed_dict={X: [2.5]}=============
[2.4999337]
=========feed_dict={X: [1.5, 3.5]}))=============
[1.5002257 3.4996414]
```

#### step = 10
```
==========step, cost_val, W_val, b_val=====
0 13.288369 [-0.29553396] [-0.4724274]
==========step, cost_val, W_val, b_val=====
10 1.2661339 [0.57624125] [-0.09125014]
==========step, cost_val, W_val, b_val=====
20 0.121346466 [0.84582084] [0.02507534]
==========step, cost_val, W_val, b_val=====
30 0.0123036755 [0.9295636] [0.05970469]
==========step, cost_val, W_val, b_val=====
40 0.0018854747 [0.95594823] [0.06915468]
==========step, cost_val, W_val, b_val=====
50 0.0008599053 [0.9646206] [0.07086428]
==========step, cost_val, W_val, b_val=====
60 0.0007302738 [0.96781486] [0.07021407]
==========step, cost_val, W_val, b_val=====
70 0.0006874558 [0.96930623] [0.0688637]
==========step, cost_val, W_val, b_val=====
80 0.0006543403 [0.9702602] [0.06732462]
==========step, cost_val, W_val, b_val=====
90 0.00062351354 [0.9710366] [0.06575398]
==========step, cost_val, W_val, b_val=====
100 0.0005942019 [0.9717466] [0.06419967]
==========step, cost_val, W_val, b_val=====
110 0.00056627695 [0.97242516] [0.06267584]
==========step, cost_val, W_val, b_val=====
120 0.0005396614 [0.97308296] [0.06118624]
==========step, cost_val, W_val, b_val=====
130 0.000514304 [0.9737237] [0.05973145]
==========step, cost_val, W_val, b_val=====
140 0.0004901315 [0.9743487] [0.05831107]
==========step, cost_val, W_val, b_val=====
150 0.0004670945 [0.9749588] [0.05692441]
==========step, cost_val, W_val, b_val=====
160 0.00044514347 [0.97555435] [0.05557071]
==========step, cost_val, W_val, b_val=====
170 0.0004242243 [0.9761357] [0.05424919]
==========step, cost_val, W_val, b_val=====
180 0.00040428768 [0.97670317] [0.0529591]
==========step, cost_val, W_val, b_val=====
190 0.00038528675 [0.9772572] [0.05169969]
==========step, cost_val, W_val, b_val=====
200 0.0003671817 [0.97779804] [0.05047024]
==========step, cost_val, W_val, b_val=====
210 0.00034992793 [0.97832596] [0.04927003]
==========step, cost_val, W_val, b_val=====
220 0.00033348193 [0.9788414] [0.04809835]
==========step, cost_val, W_val, b_val=====
230 0.00031781042 [0.9793445] [0.04695456]
==========step, cost_val, W_val, b_val=====
240 0.00030287384 [0.9798356] [0.04583799]
==========step, cost_val, W_val, b_val=====
250 0.00028863916 [0.9803154] [0.04474793]
==========step, cost_val, W_val, b_val=====
260 0.00027507325 [0.9807835] [0.04368376]
==========step, cost_val, W_val, b_val=====
270 0.00026214521 [0.98124045] [0.04264489]
==========step, cost_val, W_val, b_val=====
280 0.00024982673 [0.9816866] [0.04163074]
==========step, cost_val, W_val, b_val=====
290 0.00023808637 [0.9821221] [0.04064071]
==========step, cost_val, W_val, b_val=====
300 0.00022689575 [0.9825473] [0.03967422]
==========step, cost_val, W_val, b_val=====
310 0.00021623181 [0.9829623] [0.03873071]
==========step, cost_val, W_val, b_val=====
320 0.00020607124 [0.98336744] [0.03780967]
==========step, cost_val, W_val, b_val=====
330 0.00019638585 [0.983763] [0.03691052]
==========step, cost_val, W_val, b_val=====
340 0.00018715672 [0.98414916] [0.03603275]
==========step, cost_val, W_val, b_val=====
350 0.00017836038 [0.98452616] [0.03517585]
==========step, cost_val, W_val, b_val=====
360 0.00016997811 [0.98489416] [0.03433931]
==========step, cost_val, W_val, b_val=====
370 0.00016198978 [0.9852534] [0.03352267]
==========step, cost_val, W_val, b_val=====
380 0.00015437823 [0.98560405] [0.03272545]
==========step, cost_val, W_val, b_val=====
390 0.00014712244 [0.98594636] [0.0319472]
==========step, cost_val, W_val, b_val=====
400 0.00014020682 [0.98628056] [0.03118748]
==========step, cost_val, W_val, b_val=====
410 0.00013361806 [0.98660684] [0.03044582]
==========step, cost_val, W_val, b_val=====
420 0.00012733859 [0.98692536] [0.02972179]
==========step, cost_val, W_val, b_val=====
430 0.000121354264 [0.98723626] [0.02901497]
==========step, cost_val, W_val, b_val=====
440 0.00011564984 [0.9875398] [0.02832496]
==========step, cost_val, W_val, b_val=====
450 0.000110214685 [0.9878362] [0.02765135]
==========step, cost_val, W_val, b_val=====
460 0.00010503468 [0.9881254] [0.02699377]
==========step, cost_val, W_val, b_val=====
470 0.00010010023 [0.9884078] [0.02635184]
==========step, cost_val, W_val, b_val=====
480 9.5395946e-05 [0.98868346] [0.02572516]
==========step, cost_val, W_val, b_val=====
490 9.0911904e-05 [0.9889526] [0.02511341]
==========step, cost_val, W_val, b_val=====
500 8.664e-05 [0.9892152] [0.0245162]
==========step, cost_val, W_val, b_val=====
510 8.256786e-05 [0.9894718] [0.0239332]
==========step, cost_val, W_val, b_val=====
520 7.868841e-05 [0.98972213] [0.02336403]
==========step, cost_val, W_val, b_val=====
530 7.498992e-05 [0.98996663] [0.0228084]
==========step, cost_val, W_val, b_val=====
540 7.146429e-05 [0.9902053] [0.02226595]
==========step, cost_val, W_val, b_val=====
550 6.810623e-05 [0.99043816] [0.02173642]
==========step, cost_val, W_val, b_val=====
560 6.490495e-05 [0.99066556] [0.02121948]
==========step, cost_val, W_val, b_val=====
570 6.1854626e-05 [0.9908875] [0.02071483]
==========step, cost_val, W_val, b_val=====
580 5.894755e-05 [0.99110425] [0.0202222]
==========step, cost_val, W_val, b_val=====
590 5.6176974e-05 [0.99131584] [0.01974128]
==========step, cost_val, W_val, b_val=====
600 5.3536685e-05 [0.9915223] [0.01927179]
==========step, cost_val, W_val, b_val=====
610 5.1021078e-05 [0.9917239] [0.0188135]
==========step, cost_val, W_val, b_val=====
620 4.8622867e-05 [0.9919207] [0.0183661]
==========step, cost_val, W_val, b_val=====
630 4.633813e-05 [0.9921129] [0.01792933]
==========step, cost_val, W_val, b_val=====
640 4.4159886e-05 [0.99230045] [0.01750294]
==========step, cost_val, W_val, b_val=====
650 4.2085023e-05 [0.99248356] [0.0170867]
==========step, cost_val, W_val, b_val=====
660 4.0107276e-05 [0.9926623] [0.01668036]
==========step, cost_val, W_val, b_val=====
670 3.822266e-05 [0.9928368] [0.01628369]
==========step, cost_val, W_val, b_val=====
680 3.6425732e-05 [0.9930071] [0.01589645]
==========step, cost_val, W_val, b_val=====
690 3.4713496e-05 [0.9931734] [0.01551841]
==========step, cost_val, W_val, b_val=====
700 3.3083106e-05 [0.9933357] [0.01514938]
==========step, cost_val, W_val, b_val=====
710 3.152755e-05 [0.9934943] [0.01478911]
==========step, cost_val, W_val, b_val=====
720 3.0046698e-05 [0.99364895] [0.01443741]
==========step, cost_val, W_val, b_val=====
730 2.8634153e-05 [0.9938] [0.01409408]
==========step, cost_val, W_val, b_val=====
740 2.728796e-05 [0.99394745] [0.01375891]
==========step, cost_val, W_val, b_val=====
750 2.6006184e-05 [0.9940914] [0.01343171]
==========step, cost_val, W_val, b_val=====
760 2.4783643e-05 [0.9942319] [0.01311229]
==========step, cost_val, W_val, b_val=====
770 2.3619321e-05 [0.99436903] [0.01280047]
==========step, cost_val, W_val, b_val=====
780 2.2508806e-05 [0.9945029] [0.01249607]
==========step, cost_val, W_val, b_val=====
790 2.145161e-05 [0.9946337] [0.01219892]
==========step, cost_val, W_val, b_val=====
800 2.0442982e-05 [0.9947612] [0.01190883]
==========step, cost_val, W_val, b_val=====
810 1.9483085e-05 [0.9948858] [0.01162565]
==========step, cost_val, W_val, b_val=====
820 1.856673e-05 [0.9950074] [0.01134921]
==========step, cost_val, W_val, b_val=====
830 1.7694118e-05 [0.9951262] [0.01107933]
==========step, cost_val, W_val, b_val=====
840 1.6862628e-05 [0.9952421] [0.01081584]
==========step, cost_val, W_val, b_val=====
850 1.6070726e-05 [0.99535525] [0.01055862]
==========step, cost_val, W_val, b_val=====
860 1.5315005e-05 [0.9954657] [0.01030753]
==========step, cost_val, W_val, b_val=====
870 1.4595178e-05 [0.9955735] [0.01006241]
==========step, cost_val, W_val, b_val=====
880 1.3909636e-05 [0.9956788] [0.00982311]
==========step, cost_val, W_val, b_val=====
890 1.3256043e-05 [0.99578154] [0.00958953]
==========step, cost_val, W_val, b_val=====
900 1.263285e-05 [0.99588186] [0.00936148]
==========step, cost_val, W_val, b_val=====
910 1.2039157e-05 [0.9959798] [0.00913885]
==========step, cost_val, W_val, b_val=====
920 1.1473053e-05 [0.9960754] [0.00892153]
==========step, cost_val, W_val, b_val=====
930 1.0934064e-05 [0.99616873] [0.00870938]
==========step, cost_val, W_val, b_val=====
940 1.0420663e-05 [0.9962598] [0.00850227]
==========step, cost_val, W_val, b_val=====
950 9.930448e-06 [0.9963488] [0.00830008]
==========step, cost_val, W_val, b_val=====
960 9.464226e-06 [0.9964356] [0.00810269]
==========step, cost_val, W_val, b_val=====
970 9.01919e-06 [0.99652034] [0.00791002]
==========step, cost_val, W_val, b_val=====
980 8.595252e-06 [0.9966031] [0.00772192]
==========step, cost_val, W_val, b_val=====
990 8.191557e-06 [0.99668384] [0.0075383]
==========step, cost_val, W_val, b_val=====
1000 7.8066205e-06 [0.9967627] [0.00735905]
==========step, cost_val, W_val, b_val=====
1010 7.4395975e-06 [0.9968397] [0.00718406]
==========step, cost_val, W_val, b_val=====
1020 7.0901237e-06 [0.99691486] [0.00701323]
==========step, cost_val, W_val, b_val=====
1030 6.756822e-06 [0.99698824] [0.00684645]
==========step, cost_val, W_val, b_val=====
1040 6.4390806e-06 [0.9970598] [0.00668364]
==========step, cost_val, W_val, b_val=====
1050 6.1365567e-06 [0.99712974] [0.00652471]
==========step, cost_val, W_val, b_val=====
1060 5.8483315e-06 [0.997198] [0.00636956]
==========step, cost_val, W_val, b_val=====
1070 5.5734563e-06 [0.9972646] [0.00621811]
==========step, cost_val, W_val, b_val=====
1080 5.311618e-06 [0.9973297] [0.00607024]
==========step, cost_val, W_val, b_val=====
1090 5.0619687e-06 [0.9973932] [0.00592588]
==========step, cost_val, W_val, b_val=====
1100 4.8240404e-06 [0.9974552] [0.00578496]
==========step, cost_val, W_val, b_val=====
1110 4.5973443e-06 [0.9975157] [0.00564739]
==========step, cost_val, W_val, b_val=====
1120 4.3813016e-06 [0.99757475] [0.00551309]
==========step, cost_val, W_val, b_val=====
1130 4.1756016e-06 [0.99763244] [0.00538199]
==========step, cost_val, W_val, b_val=====
1140 3.979032e-06 [0.9976888] [0.005254]
==========step, cost_val, W_val, b_val=====
1150 3.7921354e-06 [0.9977437] [0.00512904]
==========step, cost_val, W_val, b_val=====
1160 3.6137535e-06 [0.99779737] [0.00500707]
==========step, cost_val, W_val, b_val=====
1170 3.4441534e-06 [0.99784976] [0.004888]
==========step, cost_val, W_val, b_val=====
1180 3.282186e-06 [0.9979009] [0.00477175]
==========step, cost_val, W_val, b_val=====
1190 3.1279776e-06 [0.9979508] [0.00465828]
==========step, cost_val, W_val, b_val=====
1200 2.9808982e-06 [0.9979995] [0.00454751]
==========step, cost_val, W_val, b_val=====
1210 2.8406976e-06 [0.9980471] [0.00443938]
==========step, cost_val, W_val, b_val=====
1220 2.7072429e-06 [0.99809355] [0.0043338]
==========step, cost_val, W_val, b_val=====
1230 2.5799902e-06 [0.9981389] [0.00423073]
==========step, cost_val, W_val, b_val=====
1240 2.4587478e-06 [0.99818313] [0.00413012]
==========step, cost_val, W_val, b_val=====
1250 2.3432997e-06 [0.99822634] [0.00403191]
==========step, cost_val, W_val, b_val=====
1260 2.2331108e-06 [0.9982685] [0.00393603]
==========step, cost_val, W_val, b_val=====
1270 2.1282924e-06 [0.9983097] [0.00384245]
==========step, cost_val, W_val, b_val=====
1280 2.0283114e-06 [0.9983499] [0.00375107]
==========step, cost_val, W_val, b_val=====
1290 1.932999e-06 [0.9983891] [0.00366187]
==========step, cost_val, W_val, b_val=====
1300 1.8419613e-06 [0.99842745] [0.00357479]
==========step, cost_val, W_val, b_val=====
1310 1.7556763e-06 [0.9984648] [0.00348979]
==========step, cost_val, W_val, b_val=====
1320 1.6729592e-06 [0.99850136] [0.0034068]
==========step, cost_val, W_val, b_val=====
1330 1.5943561e-06 [0.99853694] [0.00332579]
==========step, cost_val, W_val, b_val=====
1340 1.5194445e-06 [0.99857175] [0.00324671]
==========step, cost_val, W_val, b_val=====
1350 1.4480917e-06 [0.9986057] [0.00316951]
==========step, cost_val, W_val, b_val=====
1360 1.3800912e-06 [0.99863887] [0.00309413]
==========step, cost_val, W_val, b_val=====
1370 1.3151111e-06 [0.99867123] [0.00302056]
==========step, cost_val, W_val, b_val=====
1380 1.2532952e-06 [0.9987028] [0.00294873]
==========step, cost_val, W_val, b_val=====
1390 1.1944522e-06 [0.99873364] [0.00287862]
==========step, cost_val, W_val, b_val=====
1400 1.138403e-06 [0.99876374] [0.00281018]
==========step, cost_val, W_val, b_val=====
1410 1.0848365e-06 [0.9987932] [0.00274336]
==========step, cost_val, W_val, b_val=====
1420 1.0338407e-06 [0.99882185] [0.00267813]
==========step, cost_val, W_val, b_val=====
1430 9.853217e-07 [0.99884987] [0.00261446]
==========step, cost_val, W_val, b_val=====
1440 9.389278e-07 [0.99887717] [0.0025523]
==========step, cost_val, W_val, b_val=====
1450 8.950022e-07 [0.9989039] [0.00249163]
==========step, cost_val, W_val, b_val=====
1460 8.527798e-07 [0.99893] [0.00243239]
==========step, cost_val, W_val, b_val=====
1470 8.127422e-07 [0.99895537] [0.00237455]
==========step, cost_val, W_val, b_val=====
1480 7.7450903e-07 [0.9989802] [0.0023181]
==========step, cost_val, W_val, b_val=====
1490 7.381772e-07 [0.9990045] [0.00226298]
==========step, cost_val, W_val, b_val=====
1500 7.035908e-07 [0.9990281] [0.00220917]
==========step, cost_val, W_val, b_val=====
1510 6.704254e-07 [0.9990513] [0.00215666]
==========step, cost_val, W_val, b_val=====
1520 6.3886506e-07 [0.99907386] [0.00210538]
==========step, cost_val, W_val, b_val=====
1530 6.08969e-07 [0.9990959] [0.00205532]
==========step, cost_val, W_val, b_val=====
1540 5.8032634e-07 [0.9991174] [0.00200644]
==========step, cost_val, W_val, b_val=====
1550 5.530475e-07 [0.99913836] [0.00195872]
==========step, cost_val, W_val, b_val=====
1560 5.2704655e-07 [0.9991588] [0.00191214]
==========step, cost_val, W_val, b_val=====
1570 5.0230193e-07 [0.99917877] [0.00186668]
==========step, cost_val, W_val, b_val=====
1580 4.7867485e-07 [0.9991984] [0.00182231]
==========step, cost_val, W_val, b_val=====
1590 4.5620618e-07 [0.99921745] [0.00177899]
==========step, cost_val, W_val, b_val=====
1600 4.347643e-07 [0.999236] [0.00173669]
==========step, cost_val, W_val, b_val=====
1610 4.1433546e-07 [0.9992541] [0.0016954]
==========step, cost_val, W_val, b_val=====
1620 3.948202e-07 [0.9992719] [0.0016551]
==========step, cost_val, W_val, b_val=====
1630 3.764135e-07 [0.9992892] [0.00161576]
==========step, cost_val, W_val, b_val=====
1640 3.5860717e-07 [0.999306] [0.00157735]
==========step, cost_val, W_val, b_val=====
1650 3.4178166e-07 [0.9993226] [0.00153985]
==========step, cost_val, W_val, b_val=====
1660 3.2582957e-07 [0.9993387] [0.00150326]
==========step, cost_val, W_val, b_val=====
1670 3.1048324e-07 [0.99935436] [0.00146751]
==========step, cost_val, W_val, b_val=====
1680 2.9585397e-07 [0.9993698] [0.00143263]
==========step, cost_val, W_val, b_val=====
1690 2.819518e-07 [0.9993847] [0.00139858]
==========step, cost_val, W_val, b_val=====
1700 2.6871479e-07 [0.9993993] [0.00136533]
==========step, cost_val, W_val, b_val=====
1710 2.5606775e-07 [0.9994136] [0.00133289]
==========step, cost_val, W_val, b_val=====
1720 2.4409027e-07 [0.9994275] [0.00130121]
==========step, cost_val, W_val, b_val=====
1730 2.3258593e-07 [0.9994412] [0.00127029]
==========step, cost_val, W_val, b_val=====
1740 2.2163562e-07 [0.99945444] [0.00124009]
==========step, cost_val, W_val, b_val=====
1750 2.1120401e-07 [0.9994675] [0.0012106]
==========step, cost_val, W_val, b_val=====
1760 2.0133223e-07 [0.99948] [0.00118183]
==========step, cost_val, W_val, b_val=====
1770 1.9189541e-07 [0.99949247] [0.00115375]
==========step, cost_val, W_val, b_val=====
1780 1.8286438e-07 [0.9995044] [0.00112634]
==========step, cost_val, W_val, b_val=====
1790 1.7430334e-07 [0.9995163] [0.00109959]
==========step, cost_val, W_val, b_val=====
1800 1.6611837e-07 [0.9995277] [0.00107345]
==========step, cost_val, W_val, b_val=====
1810 1.583507e-07 [0.999539] [0.00104795]
==========step, cost_val, W_val, b_val=====
1820 1.5088625e-07 [0.99954987] [0.00102304]
==========step, cost_val, W_val, b_val=====
1830 1.4376111e-07 [0.9995606] [0.00099875]
==========step, cost_val, W_val, b_val=====
1840 1.3703813e-07 [0.999571] [0.000975]
==========step, cost_val, W_val, b_val=====
1850 1.306156e-07 [0.99958116] [0.00095186]
==========step, cost_val, W_val, b_val=====
1860 1.2449529e-07 [0.99959123] [0.00092924]
==========step, cost_val, W_val, b_val=====
1870 1.1867203e-07 [0.9996008] [0.00090717]
==========step, cost_val, W_val, b_val=====
1880 1.1307293e-07 [0.99961036] [0.00088563]
==========step, cost_val, W_val, b_val=====
1890 1.0774687e-07 [0.99961966] [0.00086457]
==========step, cost_val, W_val, b_val=====
1900 1.0265675e-07 [0.9996286] [0.00084404]
==========step, cost_val, W_val, b_val=====
1910 9.788929e-08 [0.99963754] [0.00082399]
==========step, cost_val, W_val, b_val=====
1920 9.3276185e-08 [0.99964607] [0.0008044]
==========step, cost_val, W_val, b_val=====
1930 8.8884484e-08 [0.9996544] [0.00078531]
==========step, cost_val, W_val, b_val=====
1940 8.47452e-08 [0.99966276] [0.00076666]
==========step, cost_val, W_val, b_val=====
1950 8.074753e-08 [0.9996707] [0.00074843]
==========step, cost_val, W_val, b_val=====
1960 7.69675e-08 [0.99967843] [0.00073068]
==========step, cost_val, W_val, b_val=====
1970 7.332803e-08 [0.9996862] [0.00071334]
==========step, cost_val, W_val, b_val=====
1980 6.9909824e-08 [0.99969363] [0.00069637]
==========step, cost_val, W_val, b_val=====
1990 6.665119e-08 [0.9997008] [0.00067985]
==========step, cost_val, W_val, b_val=====
2000 6.349562e-08 [0.99970794] [0.00066373]
=========feed_dict={X: [5]}=============
[4.9992037]
=========feed_dict={X: [2.5]}=============
[2.4999337]
=========feed_dict={X: [1.5, 3.5]}))=============
[1.5002257 3.4996414]

```

### step = 20
2000 6.349562e-08 [0.99970794] [0.00066373]

### step = 10
2000 6.349562e-08 [0.99970794] [0.00066373]

### step = 50
2000 6.349562e-08 [0.99970794] [0.00066373]

### range(3001)
3000 5.343897e-10 [0.99997306] [6.085733e-05]

### range(4001)
4000 8.436511e-12 [0.9999964] [7.521492e-06]

### range(5001)
5000 3.0695446e-12 [0.99999785] [4.3504815e-06]


### learningrate = 0.01
2000 6.349562e-08 [0.99970794] [0.00066373]

### learnignrate = 0.05
2000 3.7895614e-14 [0.99999976] [5.3116025e-07]

### learningrate = 0.5
2000 nan [nan] [nan]

### learningrate = 0.09
2000 3.7895614e-14 [0.99999976] [5.3116025e-07]

### learningrate = 0.1
**2000 0.0 [1.] [5.6859136e-08]** <br>
> 알맞는 learningrate가 있을 때, learningrate보다 큰수보다는 작은수가 좋다?

### learningrate = 0.11
2000 4.2632564e-14 [0.9999999] [9.263522e-08]

### learnignrate = 0.15
2000 3.7895614e-14 [1.0000004] [-6.212066e-07]

### learningrate = 0.25
2000 nan [nan] [nan]<br>
==========step, cost_val, W_val, b_val===== <br>
140 inf [1.6879146e+35] [7.4251644e+34] <br>
==========step, cost_val, W_val, b_val===== <br>
160 nan [nan] [nan]

### learningrate = 0.2
==========step, cost_val, W_val, b_val===== <br>
420 inf [1.9818624e+36] [8.7182496e+35] <br>
==========step, cost_val, W_val, b_val===== <br>
440 nan [nan] [nan]


In [72]:
W = tf.Variable(tf.random_normal([1]), name='weight') # Variable trainable
b = tf.Variable(tf.random_normal([1]), name='bias') # Variable

X = tf.placeholder(tf.float32, shape=[None]) 
# 그래프를 미리 만들어 놓고 실행, 그 때 값을 던져준다. 
# 특별한 노드 -> 노드값을 몰랐지만 나중에 값을 넣을 수 있다. 
Y = tf.placeholder(tf.float32, shape=[None]) # 특별한 노드

# 수식
hypo = X * W + b
# 코스트
cost = tf.reduce_mean(tf.square(hypo - Y))
# 미분 
opti = tf.train.GradientDescentOptimizer(learning_rate=0.01)
# 미분을 사용해서? cost function을 최소화?
train = opti.minimize(cost)

sess = tf.Session()
sess.run(tf.global_variables_initializer())

for step in range(2001):
    cost_val, W_val, b_val, _ = sess.run([cost, W, b, train],
                                        feed_dict={X: [1, 2, 3, 4, 5],
                                        Y: [2.1, 3.1, 4.1, 5.1, 6.1]})
    if step % 20 == 0:
        print(step, cost_val, W_val, b_val)

0 15.623711 [0.3858341] [0.04400694]
20 0.10865022 [1.2091624] [0.33007258]
40 0.094608955 [1.1990013] [0.38147542]
60 0.0826227 [1.1859845] [0.4285366]
80 0.07215501 [1.1738042] [0.47251126]
100 0.06301346 [1.1624217] [0.51360613]
120 0.0550301 [1.1517845] [0.5520096]
140 0.04805824 [1.141844] [0.58789796]
160 0.041969582 [1.1325545] [0.62143594]
180 0.03665236 [1.1238735] [0.65277755]
200 0.03200877 [1.1157608] [0.6820666]
220 0.02795344 [1.1081796] [0.7094374]
240 0.024411976 [1.1010947] [0.73501563]
260 0.021319162 [1.094474] [0.7589188]
280 0.018618181 [1.0882868] [0.78125656]
300 0.016259387 [1.0825049] [0.80213124]
320 0.014199448 [1.0771016] [0.8216388]
340 0.0124005 [1.0720521] [0.83986884]
360 0.010829445 [1.0673333] [0.8569051]
380 0.00945742 [1.0629236] [0.8728258]
400 0.00825922 [1.0588027] [0.88770366]
420 0.007212843 [1.0549517] [0.90160716]
440 0.0062990314 [1.0513529] [0.9146]
460 0.0055009765 [1.0479896] [0.9267421]
480 0.004804056 [1.0448468] [0.93808883]
500 0.00419

#### 대략 W = 1, b=1.1 
X : [1, 2, 3, 4, 5] Y: [2.1, 3.1, 4.1, 5.1, 6.1] <br>
1 \* 1 + 1.1 = 2.1 <br>
2 \* 1 + 1.1 = 3.1 <br>

### 학습한 값이 잘 수행되는지 확인

In [73]:
print(sess.run(hypo, feed_dict={X: [5]}))
print(sess.run(hypo, feed_dict={X: [2.5]}))
print(sess.run(hypo, 
               feed_dict={X: [1.5, 3.5]}))

[6.100363]
[3.5997102]
[2.5994494 4.5999713]


In [78]:
print(5*1+1.1)
print(2.5*1+1.1)

6.1
3.6


In [88]:
a = [1.5, 3.5]
print(a * 1)
for i in a:
    print(i + 1.1)
    
print("=======index로 접근=======")    
for idx in range(len(a)):
    print(idx)
    print(a[idx])
    print(a[idx]+1.1)
    a[idx] = a[idx]+1.1
    
print("List 안의 숫자들은")
print(a)

[1.5, 3.5]
2.6
4.6
=======index로 접근=======
0
1.5
2.6
1
3.5
4.6
List 안의 숫자들은
[2.6, 4.6]
